### ISSUE TITLE
Use Poolable for CurvePoint

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/calculus-grapher#320]
While working on #309 and #312, @zepumph recommended using Poolable for CurvePoint, to reduce the amount of garbage collection required when setting state.  This will be a major change, so creating this issue to track.

**Comment by pixelzoom** [comment]
@zepumph @veillette and I got partway through this, but it was still failing in the State Wrapper.  We'll need help from @zepumph to complete this.  Here's a patch of where we left off.  Assigning to @zepumph, let me know if you'd like me to pair on Zoom, or if you'd prefer to move it forward yourself before we reconvene.

<details>
<summary>patch</summary>

```diff
Subject: [PATCH] clipboardPatchFile
---
Index: js/common/model/CurvePoint.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/model/CurvePoint.ts b/js/common/model/CurvePoint.ts
--- a/js/common/model/CurvePoint.ts	(revision c61ca56501979750b0a7f2ee4049a2398b2f2cd0)
+++ b/js/common/model/CurvePoint.ts	(date 1680017284198)
@@ -17,11 +17,13 @@
  */
 
 import Vector2 from '../../../../dot/js/Vector2.js';
+import Pool, { TPoolable } from '../../../../phet-core/js/Pool.js';
 import IOType from '../../../../tandem/js/types/IOType.js';
 import NumberIO from '../../../../tandem/js/types/NumberIO.js';
 import StringUnionIO from '../../../../tandem/js/types/StringUnionIO.js';
 import calculusGrapher from '../../calculusGrapher.js';
 import CalculusGrapherConstants from '../CalculusGrapherConstants.js';
+import CalculusGrapherQueryParameters from '../CalculusGrapherQueryParameters.js';
 
 const PointTypeValues = [ 'smooth', 'cusp', 'discontinuous' ] as const;
 export type PointType = ( typeof PointTypeValues )[number];
@@ -41,10 +43,10 @@
   initialPointType: PointType;
 };
 
-export default class CurvePoint {
+export default class CurvePoint implements TPoolable {
 
   // The x coordinate is fixed.
-  public readonly x: number;
+  public x: number; //TODO https://github.com/phetsims/calculus-grapher/issues/310 should be readonly, change to _x and add a getter
 
   // Using an observable Property for the y-value was considered, but it was deemed to be
   // invasive to the performance of the simulation as observers had to listen to the yProperty
@@ -69,7 +71,7 @@
   // at instantiation, then reused to prevent garbage collection.
   private readonly vector: Vector2;
 
-  public constructor( x: number, y = 0, pointType: PointType = 'smooth' ) {
+  public constructor( x: number, y: number, pointType: PointType = 'smooth' ) {
     assert && assert( Number.isFinite( x ), `invalid x: ${x}` );
 
     this.x = x;
@@ -85,6 +87,27 @@
     this.vector = new Vector2( x, y );
   }
 
+  public static readonly pool = new Pool( CurvePoint, {
+    maxSize: CalculusGrapherQueryParameters.numberOfPoints, //TODO https://github.com/phetsims/calculus-grapher/issues/320
+    initialize: CurvePoint.prototype.initialize,
+    defaultArguments: [ 0, 0, 'smooth' ]
+  } );
+
+  public freeToPool(): void {
+    CurvePoint.pool.freeToPool( this );
+  }
+
+  public initialize( x: number, y:number, pointType: PointType = 'smooth' ): CurvePoint {
+    this.x = x;
+    this.y = y;
+    this.pointType = pointType;
+    this.initialState.y = y;
+    this.initialState.pointType = pointType;
+    this.undoStack.length = 0;
+    this.vector.setXY( x, y );
+    return this;
+  }
+
   /**
    * Returns a reference to the Vector2 representation of this CurvePoint. DO NOT MODIFY!
    */
@@ -196,7 +219,7 @@
   }
 
   public static fromStateObject( stateObject: CurvePointStateObject ): CurvePoint {
-    const curvePoint = new CurvePoint(
+    const curvePoint = CurvePoint.pool.create(
       stateObject.x,
       stateObject.initialY,
       stateObject.initialPointType
Index: js/common/model/Curve.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/model/Curve.ts b/js/common/model/Curve.ts
--- a/js/common/model/Curve.ts	(revision c61ca56501979750b0a7f2ee4049a2398b2f2cd0)
+++ b/js/common/model/Curve.ts	(date 1680017104759)
@@ -27,12 +27,14 @@
 import calculusGrapher from '../../calculusGrapher.js';
 import CalculusGrapherConstants from '../CalculusGrapherConstants.js';
 import CalculusGrapherQueryParameters from '../CalculusGrapherQueryParameters.js';
-import CurvePoint from './CurvePoint.js';
+import CurvePoint, { CurvePointStateObject } from './CurvePoint.js';
 import Range from '../../../../dot/js/Range.js';
 import optionize from '../../../../phet-core/js/optionize.js';
 import Property from '../../../../axon/js/Property.js';
 import ArrayIO from '../../../../tandem/js/types/ArrayIO.js';
 import Tandem from '../../../../tandem/js/Tandem.js';
+import IOType from '../../../../tandem/js/types/IOType.js';
+import ReadOnlyProperty from '../../../../axon/js/ReadOnlyProperty.js';
 
 // Constants
 const CURVE_X_RANGE = CalculusGrapherConstants.CURVE_X_RANGE;
@@ -107,13 +109,25 @@
     for ( let i = 0; i < this.numberOfPoints; i++ ) {
       const xNormalized = i / ( this.numberOfPoints - 1 ); // in the range [0,1]
       const x = this.xRange.expandNormalizedValue( xNormalized );
-      initialPoints.push( new CurvePoint( x, 0 ) );
+      initialPoints.push( CurvePoint.pool.create( x, 0 ) );
     }
 
     this.pointsProperty = new Property( initialPoints, {
       isValidValue: points => isValidPoints( initialPoints, points ),
       tandem: options.tandem.createTandem( 'pointsProperty' ),
       phetioValueType: ArrayIO( CurvePoint.CurvePointIO ),
+
+      //TODO https://github.com/phetsims/calculus-grapher/issues/320 more documentation
+      phetioOuterType: parameterType => {
+        return new IOType<Property<Array<CurvePoint>>, CurvePointStateObject[]>( 'PointerPropertyIO', {
+          supertype: ReadOnlyProperty.PropertyIO( parameterType ),
+          applyState: ( property, stateObject ) => {
+            property.value.forEach( point => point.freeToPool() ); // free CurvePoints to pool
+            ReadOnlyProperty.PropertyIO( parameterType ).applyState( property, stateObject );
+          },
+          parameterTypes: [ parameterType ]
+        } );
+      },
       phetioReadOnly: options.pointsPropertyReadOnly,
       phetioFeatured: true,
       phetioDocumentation: options.pointsPropertyReadOnly ? DOC_READ_ONLY_TRUE : DOC_READ_ONLY_FALSE
@@ -136,6 +150,7 @@
     if ( Tandem.PHET_IO_ENABLED && this.pointsProperty.isPhetioInstrumented() && !this.pointsProperty.phetioReadOnly ) {
       this.pointsProperty.lazyLink( () => {
         phet.log && phet.log( `pointsProperty changed: ${this.phetioID}` );
+        console.log( `CurvePoint poolSize = ${CurvePoint.pool.poolSize}` ); //TODO https://github.com/phetsims/calculus-grapher/issues/320 delete or use phet.log
         this.curveChangedEmitter.emit();
       } );
     }
```

</details>

**Comment by zepumph** [comment]
We fixed the bug with validation, and then found with testing that freeing to pool seemed a bit slow. My state wrapper was a bit laggy, I think it was just my computer, but @pixelzoom should test also to see.

<details>

```diff
Subject: [PATCH] update doc with full recording link, https://github.com/phetsims/quadrilateral/issues/97
---
Index: calculus-grapher/js/common/model/CurvePoint.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/calculus-grapher/js/common/model/CurvePoint.ts b/calculus-grapher/js/common/model/CurvePoint.ts
--- a/calculus-grapher/js/common/model/CurvePoint.ts	(revision c61ca56501979750b0a7f2ee4049a2398b2f2cd0)
+++ b/calculus-grapher/js/common/model/CurvePoint.ts	(date 1680019353760)
@@ -17,11 +17,13 @@
  */
 
 import Vector2 from '../../../../dot/js/Vector2.js';
+import Pool, { TPoolable } from '../../../../phet-core/js/Pool.js';
 import IOType from '../../../../tandem/js/types/IOType.js';
 import NumberIO from '../../../../tandem/js/types/NumberIO.js';
 import StringUnionIO from '../../../../tandem/js/types/StringUnionIO.js';
 import calculusGrapher from '../../calculusGrapher.js';
 import CalculusGrapherConstants from '../CalculusGrapherConstants.js';
+import CalculusGrapherQueryParameters from '../CalculusGrapherQueryParameters.js';
 
 const PointTypeValues = [ 'smooth', 'cusp', 'discontinuous' ] as const;
 export type PointType = ( typeof PointTypeValues )[number];
@@ -41,10 +43,10 @@
   initialPointType: PointType;
 };
 
-export default class CurvePoint {
+export default class CurvePoint implements TPoolable {
 
   // The x coordinate is fixed.
-  public readonly x: number;
+  public x: number; //TODO https://github.com/phetsims/calculus-grapher/issues/310 should be readonly, change to _x and add a getter
 
   // Using an observable Property for the y-value was considered, but it was deemed to be
   // invasive to the performance of the simulation as observers had to listen to the yProperty
@@ -69,7 +71,7 @@
   // at instantiation, then reused to prevent garbage collection.
   private readonly vector: Vector2;
 
-  public constructor( x: number, y = 0, pointType: PointType = 'smooth' ) {
+  public constructor( x: number, y: number, pointType: PointType = 'smooth' ) {
     assert && assert( Number.isFinite( x ), `invalid x: ${x}` );
 
     this.x = x;
@@ -85,6 +87,27 @@
     this.vector = new Vector2( x, y );
   }
 
+  public static readonly pool = new Pool( CurvePoint, {
+    maxSize: CalculusGrapherQueryParameters.numberOfPoints, //TODO https://github.com/phetsims/calculus-grapher/issues/320
+    initialize: CurvePoint.prototype.initialize,
+    defaultArguments: [ 0, 0, 'smooth' ]
+  } );
+
+  public freeToPool(): void {
+    CurvePoint.pool.freeToPool( this );
+  }
+
+  public initialize( x: number, y:number, pointType: PointType = 'smooth' ): CurvePoint {
+    this.x = x;
+    this.y = y;
+    this.pointType = pointType;
+    this.initialState.y = y;
+    this.initialState.pointType = pointType;
+    this.undoStack.length = 0;
+    this.vector.setXY( x, y );
+    return this;
+  }
+
   /**
    * Returns a reference to the Vector2 representation of this CurvePoint. DO NOT MODIFY!
    */
@@ -196,7 +219,7 @@
   }
 
   public static fromStateObject( stateObject: CurvePointStateObject ): CurvePoint {
-    const curvePoint = new CurvePoint(
+    const curvePoint = CurvePoint.pool.create(
       stateObject.x,
       stateObject.initialY,
       stateObject.initialPointType
Index: phet-core/js/Pool.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/phet-core/js/Pool.ts b/phet-core/js/Pool.ts
--- a/phet-core/js/Pool.ts	(revision c0a87d344680f0806a10d9508e73c00742a6c4ed)
+++ b/phet-core/js/Pool.ts	(date 1680019353471)
@@ -11,11 +11,11 @@
  *
  * With this style of pooling, the following should be standard boilerplate within the class:
 
-  public freeToPool(): void {
+ public freeToPool(): void {
     MyType.pool.freeToPool( this );
   }
 
-  public static readonly pool = new Pool( MyType );
+ public static readonly pool = new Pool( MyType );
 
  * and can additionally implement TPoolable to make it clear that the type is pooled
  *
@@ -50,9 +50,9 @@
 
 export type PoolableOptions<T extends Constructor> =
   SelfPoolableOptions<T> & ( InstanceType<T> extends { initialize: PoolableInitializer<T> } ? unknown : {
-    // Require initialize if our type doesn't have a compatible initialize method.
-    initialize: PoolableInitializer<T>;
-  } );
+  // Require initialize if our type doesn't have a compatible initialize method.
+  initialize: PoolableInitializer<T>;
+} );
 type PoolableInitializer<T extends Constructor> = ( ...args: ConstructorParameters<T> ) => InstanceType<T>;
 
 export type TPoolable = {
@@ -175,6 +175,11 @@
     }
   }
 
+  public freeManyToPool( objects: InstanceType<T>[] ): void {
+    const numberToAdd = Math.min( objects.length, this.maxPoolSize - this.objects.length );
+    this.objects.push.apply( this.objects, objects.slice( 0, numberToAdd ) ); // eslint-disable-line
+  }
+
   public forEach( callback: ( object: InstanceType<T> ) => void ): void {
     this.objects.forEach( callback );
   }
Index: calculus-grapher/js/common/model/Curve.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/calculus-grapher/js/common/model/Curve.ts b/calculus-grapher/js/common/model/Curve.ts
--- a/calculus-grapher/js/common/model/Curve.ts	(revision c61ca56501979750b0a7f2ee4049a2398b2f2cd0)
+++ b/calculus-grapher/js/common/model/Curve.ts	(date 1680019499958)
@@ -27,12 +27,14 @@
 import calculusGrapher from '../../calculusGrapher.js';
 import CalculusGrapherConstants from '../CalculusGrapherConstants.js';
 import CalculusGrapherQueryParameters from '../CalculusGrapherQueryParameters.js';
-import CurvePoint from './CurvePoint.js';
+import CurvePoint, { CurvePointStateObject } from './CurvePoint.js';
 import Range from '../../../../dot/js/Range.js';
 import optionize from '../../../../phet-core/js/optionize.js';
 import Property from '../../../../axon/js/Property.js';
 import ArrayIO from '../../../../tandem/js/types/ArrayIO.js';
 import Tandem from '../../../../tandem/js/Tandem.js';
+import IOType from '../../../../tandem/js/types/IOType.js';
+import ReadOnlyProperty from '../../../../axon/js/ReadOnlyProperty.js';
 
 // Constants
 const CURVE_X_RANGE = CalculusGrapherConstants.CURVE_X_RANGE;
@@ -107,13 +109,32 @@
     for ( let i = 0; i < this.numberOfPoints; i++ ) {
       const xNormalized = i / ( this.numberOfPoints - 1 ); // in the range [0,1]
       const x = this.xRange.expandNormalizedValue( xNormalized );
-      initialPoints.push( new CurvePoint( x, 0 ) );
+      initialPoints.push( CurvePoint.pool.create( x, 0 ) );
     }
 
     this.pointsProperty = new Property( initialPoints, {
-      isValidValue: points => isValidPoints( initialPoints, points ),
+      // isValidValue: points => isValidPoints( initialPoints, points ),
       tandem: options.tandem.createTandem( 'pointsProperty' ),
       phetioValueType: ArrayIO( CurvePoint.CurvePointIO ),
+
+      //TODO https://github.com/phetsims/calculus-grapher/issues/320 more documentation
+      phetioOuterType: parameterType => {
+        return new IOType<Property<Array<CurvePoint>>, CurvePointStateObject[]>( 'PointerPropertyIO', {
+          supertype: ReadOnlyProperty.PropertyIO( parameterType ),
+          applyState: ( property, stateObject ) => {
+            console.log( `CurvePoint poolSize pre free = ${CurvePoint.pool.poolSize}` ); //TODO https://github.com/phetsims/calculus-grapher/issues/320 delete or use phet.log
+
+            const start = Date.now();
+            CurvePoint.pool.freeManyToPool( property.value ); // free CurvePoints to pool
+            console.log( Date.now() - start, 'time to free' );
+            console.log( `CurvePoint poolSize post free = ${CurvePoint.pool.poolSize}` ); //TODO https://github.com/phetsims/calculus-grapher/issues/320 delete or use phet.log
+            ReadOnlyProperty.PropertyIO( parameterType ).applyState( property, stateObject );
+            console.log( `CurvePoint poolSize post apply = ${CurvePoint.pool.poolSize}` ); //TODO https://github.com/phetsims/calculus-grapher/issues/320 delete or use phet.log
+
+          },
+          parameterTypes: [ parameterType ]
+        } );
+      },
       phetioReadOnly: options.pointsPropertyReadOnly,
       phetioFeatured: true,
       phetioDocumentation: options.pointsPropertyReadOnly ? DOC_READ_ONLY_TRUE : DOC_READ_ONLY_FALSE
@@ -136,6 +157,7 @@
     if ( Tandem.PHET_IO_ENABLED && this.pointsProperty.isPhetioInstrumented() && !this.pointsProperty.phetioReadOnly ) {
       this.pointsProperty.lazyLink( () => {
         phet.log && phet.log( `pointsProperty changed: ${this.phetioID}` );
+        console.log( `CurvePoint poolSize = ${CurvePoint.pool.poolSize}` ); //TODO https://github.com/phetsims/calculus-grapher/issues/320 delete or use phet.log
         this.curveChangedEmitter.emit();
       } );
     }
@@ -188,13 +210,7 @@
 function isValidPoints( initialPoints: CurvePoint[], points: CurvePoint[] ): boolean {
 
   // The number of points must be the same.
-  let isValid = ( initialPoints.length === points.length );
-
-  // All x coordinates must be the same.
-  for ( let i = 0; i < initialPoints.length && isValid; i++ ) {
-    isValid = ( initialPoints[ i ].x === points[ i ].x );
-  }
-  return isValid;
+  return initialPoints.length === points.length;
 }
 
 calculusGrapher.register( 'Curve', Curve );


```
</details>

**Comment by pixelzoom** [comment]
@zepumph and I worked on this some more, using the patch in https://github.com/phetsims/calculus-grapher/issues/320#issuecomment-1487210823.

With Poolable, the time to set state (in the State Wrapper) increased about 25% (from 1700 ms to 2150 ms on MacBook Pro).   Doing a Performance recording without Poolable showed little impact due to GC.  And most importantly, doing a Performance recording _with_ Poolable suggested that we have a serious (new) memory leak - and now is not the time to investigate.

So since none of this is likely to impact the product that clients deliver, we decided to punt.  Closing this issue.
